File: java/structures/SegmentTree.java
package structures;

import java.util.function.Predicate;

public class SegmentTree {
    int n;
    Node[] tree;

    public static class Node {
        // initial values for leaves
        public long mx = 0;
        public long sum = 0;
        public long add = 0;

        void apply(int l, int r, long v) {
            mx += v;
            sum += v * (r - l + 1);
            add += v;
        }
    }

    static Node unite(Node a, Node b) {
        Node res = new Node();
        res.mx = Math.max(a.mx, b.mx);
        res.sum = a.sum + b.sum;
        return res;
    }

    void push(int x, int l, int r) {
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        if (tree[x].add != 0) {
            tree[x + 1].apply(l, m, tree[x].add);
            tree[y].apply(m + 1, r, tree[x].add);
            tree[x].add = 0;
        }
    }

    void pull(int x, int y) {
        tree[x] = unite(tree[x + 1], tree[y]);
    }

    SegmentTree(int n) {
        this.n = n;
        tree = new Node[2 * n - 1];
        for (int i = 0; i < tree.length; i++) tree[i] = new Node();
        build(0, 0, n - 1);
    }

    SegmentTree(long[] v) {
        n = v.length;
        tree = new Node[2 * n - 1];
        for (int i = 0; i < tree.length; i++) tree[i] = new Node();
        build(0, 0, n - 1, v);
    }

    void build(int x, int l, int r) {
        if (l == r) {
            return;
        }
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        build(x + 1, l, m);
        build(y, m + 1, r);
        pull(x, y);
    }

    void build(int x, int l, int r, long[] v) {
        if (l == r) {
            tree[x].apply(l, r, v[l]);
            return;
        }
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        build(x + 1, l, m, v);
        build(y, m + 1, r, v);
        pull(x, y);
    }

    public Node get(int ll, int rr) {
        return get(ll, rr, 0, 0, n - 1);
    }

    Node get(int ll, int rr, int x, int l, int r) {
        if (ll <= l && r <= rr) {
            return tree[x];
        }
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        push(x, l, r);
        Node res;
        if (rr <= m) {
            res = get(ll, rr, x + 1, l, m);
        } else {
            if (ll > m) {
                res = get(ll, rr, y, m + 1, r);
            } else {
                res = unite(get(ll, rr, x + 1, l, m), get(ll, rr, y, m + 1, r));
            }
        }
        pull(x, y);
        return res;
    }

    void modify(int ll, int rr, long v) {
        modify(ll, rr, v, 0, 0, n - 1);
    }

    void modify(int ll, int rr, long v, int x, int l, int r) {
        if (ll <= l && r <= rr) {
            tree[x].apply(l, r, v);
            return;
        }
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        push(x, l, r);
        if (ll <= m) {
            modify(ll, rr, v, x + 1, l, m);
        }
        if (rr > m) {
            modify(ll, rr, v, y, m + 1, r);
        }
        pull(x, y);
    }

    // calls all FALSE elements to the left of the sought position exactly once
    int findFirst(int ll, int rr, Predicate<Node> f) {
        return findFirst(ll, rr, f, 0, 0, n - 1);
    }

    int findFirst(int ll, int rr, Predicate<Node> f, int x, int l, int r) {
        if (ll <= l && r <= rr && !f.test(tree[x])) {
            return -1;
        }
        if (l == r) {
            return l;
        }
        push(x, l, r);
        int m = (l + r) >> 1;
        int y = x + ((m - l + 1) << 1);
        int res = -1;
        if (ll <= m) {
            res = findFirst(ll, rr, f, x + 1, l, m);
        }
        if (rr > m && res == -1) {
            res = findFirst(ll, rr, f, y, m + 1, r);
        }
        pull(x, y);
        return res;
    }

    // Returns min(p | p<=rr && sum[ll..p]>=sum). If no such p exists, returns -1
    static int sumLowerBound(SegmentTree t, int ll, int rr, long sum) {
        long[] sumSoFar = new long[1];
        return t.findFirst(ll, rr, node -> {
            if (sumSoFar[0] + node.sum >= sum)
                return true;
            sumSoFar[0] += node.sum;
            return false;
        });
    }

    // Usage example
    public static void main(String[] args) {
        SegmentTree t = new SegmentTree(10);
        t.modify(1, 2, 10);
        t.modify(2, 3, 20);
        System.out.println(30 == t.get(1, 3).mx);
        System.out.println(60 == t.get(1, 3).sum);

        SegmentTree tt = new SegmentTree(new long[] {2, 1, 10, 20});
        System.out.println(2 == sumLowerBound(tt, 0, tt.n - 1, 12));
    }
}

File: golang/structures/segment_tree.go
package structures

// SegmentTree with lazy propagation (Add/Max operations)
type SegmentTree struct {
	n     int
	value []int
	delta []int
}

func NewSegmentTree(n int) *SegmentTree {
	return &SegmentTree{
		n:     n,
		value: make([]int, 4*n),
		delta: make([]int, 4*n),
	}
}

func (t *SegmentTree) pushDelta(root, left, right int) {
	t.value[root] += t.delta[root]
	t.delta[2*root+1] += t.delta[root]
	t.delta[2*root+2] += t.delta[root]
	t.delta[root] = 0
}

func (t *SegmentTree) Query(from, to int) int {
	return t.query(from, to, 0, 0, t.n-1)
}

func (t *SegmentTree) query(from, to, root, left, right int) int {
	if from == left && to == right {
		return t.value[root] + t.delta[root]
	}
	t.pushDelta(root, left, right)
	mid := (left + right) >> 1
	if from <= mid && to > mid {
		l := t.query(from, min(to, mid), root*2+1, left, mid)
		r := t.query(max(from, mid+1), to, root*2+2, mid+1, right)
		if l > r {
			return l
		}
		return r
	} else if from <= mid {
		return t.query(from, min(to, mid), root*2+1, left, mid)
	}
	return t.query(max(from, mid+1), to, root*2+2, mid+1, right)
}

func (t *SegmentTree) Modify(from, to, delta int) {
	t.modify(from, to, delta, 0, 0, t.n-1)
}

func (t *SegmentTree) modify(from, to, delta, root, left, right int) {
	if from == left && to == right {
		t.delta[root] += delta
		return
	}
	t.pushDelta(root, left, right)
	mid := (left + right) >> 1
	if from <= mid {
		t.modify(from, min(to, mid), delta, 2*root+1, left, mid)
	}
	if to > mid {
		t.modify(max(from, mid+1), to, delta, 2*root+2, mid+1, right)
	}
	lv := t.value[2*root+1] + t.delta[2*root+1]
	rv := t.value[2*root+2] + t.delta[2*root+2]
	if lv > rv {
		t.value[root] = lv
	} else {
		t.value[root] = rv
	}
}
